lib/usb.ts

```diff
@@ -460,21 +476,81 @@ class MicrobitWebUSBConnectionImpl
   }
 
   private async connectInternal(): Promise<void> {
-    if (!this.connection) {
-      const device = await this.chooseDevice();
-      this.connection = new DAPWrapper(device, this.logging);
+    if (!this.connection && this.device) {
+      this.connection = new DAPWrapper(this.device, this.logging);
+      await withTimeout(this.connection.reconnectAsync(), 10_000);
+    } else if (!this.connection) {
+      await this.connectWithPreviouslyPairedOrSelectedDevice();
+    } else {
+      await withTimeout(this.connection.reconnectAsync(), 10_000);
     }
-    await withTimeout(this.connection.reconnectAsync(), 10_000);
     if (this.addedListeners.serialdata && !this.flashing) {
       this.startSerialInternal();
     }
     this.setStatus(ConnectionStatus.CONNECTED);
   }
 
-  private async chooseDevice(): Promise<USBDevice> {
-    if (this.device) {
-      return this.device;
+  private async connectWithPreviouslyPairedOrSelectedDevice(): Promise<void> {
+    if (this.deviceConnectMode === DeviceConnectMode.StoredOrKnownOrSelect) {
+      await this.tryPreviouslyPairedDevices();
+    }
+    if (!this.connection) {
+      this.device = await this.chooseDevice();
+      this.connection = new DAPWrapper(this.device, this.logging);
+      await withTimeout(this.connection.reconnectAsync(), 10_000);
+    }
+  }
+
+  /**
+   * Attempts to connect to previously paired devices
+   * Based on: https://github.com/microsoft/pxt/blob/ab97a2422879824c730f009b15d4bf446b0e8547/pxtlib/webusb.ts#L361
+   */
+  private async tryPreviouslyPairedDevices(): Promise<void> {
+    const pairedDevices = await this.getPreviouslyPairedDevices();
+    for (const device of pairedDevices) {
+      if (await this.attemptDeviceConnection(device)) {
+        return; // Successfully connected
+      }
     }
+  }
+
+  /**
+   * Retrieves previously paired USB devices
+   * Based on: https://github.com/microsoft/pxt/blob/ab97a2422879824c730f009b15d4bf446b0e8547/pxtlib/webusb.ts#L530
+   */
+  private async getPreviouslyPairedDevices(): Promise<USBDevice[]> {
+    this.log("Retrieving previously paired USB devices");
+    try {
+      const devices = await this.withEnrichedErrors(() =>
+        navigator.usb?.getDevices(),
+      );
+      return devices ?? [];
+    } catch (error: any) {
+      this.log(`Failed to retrieve paired devices: ${error.message}`);
+      return [];
+    }
+  }
+
+  private async attemptDeviceConnection(device: USBDevice): Promise<boolean> {
+    this.device = device;
+    this.log(
+      `Attempting connection to: ${device.manufacturerName} ${device.productName}`,
+    );
+    this.log(`Serial number: ${device.serialNumber}`);
+    try {
+      this.connection = new DAPWrapper(device, this.logging);
```

Comment:
Do we need to dispose / clean up the DAPWrapper?

Perhaps this code would be cleaner if it uses locals until successful? Or do we need to assign to fields as we go?